Loop/Managers/RemoteDataServicesManager.swift

```diff
@@ -6,165 +6,362 @@
-        remoteDataServices.removeAll { $0.serviceIdentifier == remoteDataService.serviceIdentifier }
+        lock.withLock {
+            unlockedRemoteDataServices.removeAll { $0.serviceIdentifier == remoteDataService.serviceIdentifier }
+        }
+        clearQueryAnchors(for: remoteDataService)
     }
 
-    @objc func loopDataUpdated(_ note: Notification) {
-        guard
-            !remoteDataServices.isEmpty,
-            let rawContext = note.userInfo?[LoopDataManager.LoopUpdateContextKey] as? LoopDataManager.LoopUpdateContext.RawValue,
-            let context = LoopDataManager.LoopUpdateContext(rawValue: rawContext),
-            case .preferences = context
-            else {
-                return
+    private var remoteDataServices: [RemoteDataService] { return lock.withLock { unlockedRemoteDataServices } }
+
+    private func dispatchQueue(for remoteDataService: RemoteDataService, withDataType dataType: String) -> DispatchQueue {
+        return lock.withLock {
+            let dispatchQueueName = self.dispatchQueueName(for: remoteDataService, withDataType: dataType)
+
+            if let dispatchQueue = self.unlockedDispatchQueues[dispatchQueueName] {
+                return dispatchQueue
+            }
+
+            let dispatchQueue = DispatchQueue(label: dispatchQueueName, qos: .utility)
+            self.unlockedDispatchQueues[dispatchQueueName] = dispatchQueue
+            return dispatchQueue
         }
+    }
 
-        lastSettingsUpdate = Date()
+    private func dispatchQueueName(for remoteDataService: RemoteDataService, withDataType dataType: String) -> String {
+        return "com.loopkit.Loop.RemoteDataServicesManager.\(remoteDataService.serviceIdentifier).\(dataType)DispatchQueue"
+    }
 
-        uploadSettings()
+    private func clearQueryAnchors(for remoteDataService: RemoteDataService) {
+        clearCarbQueryAnchor(for: remoteDataService)
+        clearDoseQueryAnchor(for: remoteDataService)
+        clearGlucoseQueryAnchor(for: remoteDataService)
+        clearPumpEventQueryAnchor(for: remoteDataService)
+        clearSettingsQueryAnchor(for: remoteDataService)
+        clearStatusQueryAnchor(for: remoteDataService)
     }
 
-    private func uploadSettings() {
-        guard !remoteDataServices.isEmpty else {
-            return
+}
+
+extension RemoteDataServicesManager {
+
+    private var carbDataType: String { return "Carb" }
+
+    public func carbStoreHasUpdatedCarbData(_ carbStore: CarbStore) {
+        remoteDataServices.forEach { self.synchronizeCarbData(from: carbStore, to: $0) }
+    }
+
+    private func synchronizeCarbData(from carbStore: CarbStore, to remoteDataService: RemoteDataService) {
+        dispatchQueue(for: remoteDataService, withDataType: carbDataType).async {
+            let semaphore = DispatchSemaphore(value: 0)
+            let queryAnchor = UserDefaults.appGroup?.getQueryAnchor(for: remoteDataService, withDataType: self.carbDataType) ?? CarbStore.QueryAnchor()
```

Comment:
I hate to be adding more dependencies to UserDefaults at this point. But we can keep the migration of this to a different store for a separate task.

Reply:
Perhaps add a Jira ticket to track that we need to implement a store for query anchors.

Reply:
I will create a new JIRA story.

Reply:
Done. https://tidepool.atlassian.net/browse/LOOP-348